### ISSUE TITLE
AlreadySpiedError not being thrown

### ISSUE THREAD

**Issue by jbodah** [jbodah/spy_rb#16]
```rb
irb(main):001:0> require 'spy'
=> true
irb(main):002:0> Spy.on(Object, :to_s)
=> #<Spy::Instance:0x007fd8e93e4da0 @spied=Object, @original=#<Method: Class(Module)#to_s>, @visibility=:public, @conditional_filters=[], @before_callbacks=[], @after_callbacks=[], @around_procs=[], @call_history=[], @strategy=#<Spy::Instance::Strategy::Intercept:0x007fd8e93e4af8 @spy=#<Spy::Instance:0x007fd8e93e4da0 ...>, @intercept_target=#<Class:Object>>>
irb(main):003:0> Spy.on(Object, :to_s)
=> #<Spy::Instance:0x007fd8e93dea40 @spied=Object, @original=#<Method: Object.to_s>, @visibility=:public, @conditional_filters=[], @before_callbacks=[], @after_callbacks=[], @around_procs=[], @call_history=[], @strategy=#<Spy::Instance::Strategy::Wrap:0x007fd8e93de8b0 @spy=#<Spy::Instance:0x007fd8e93dea40 ...>>>
irb(main):004:0> Spy.on(Object, :to_s)
Object: Spy::Errors::AlreadySpiedError
        from /Users/Bodah/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/spy_rb-0.3.0/lib/spy/collection.rb:12:in `insert'
        from /Users/Bodah/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/spy_rb-0.3.0/lib/spy/collection.rb:39:in `<<'
        from /Users/Bodah/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/spy_rb-0.3.0/lib/spy/core.rb:13:in `add_spy'
        from /Users/Bodah/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/spy_rb-0.3.0/lib/spy/api.rb:10:in `on'
        from (irb):4
        from /Users/Bodah/.rbenv/versions/2.1.3/bin/irb:11:in `<main>'
```

**Comment by jbodah** [comment]
The issue is likely:

1. Spy called with to_s; Module is the owner of to_s
2. Spy::Collection logs the spy
3. Intercept strategy is used defining the method on the singleton
4. Spy called with to_s; Singleton is the owner of to_s
5. Spy::Collection doesn't find that the spy is logged
6. Wrap strategy is used redefining the method on the singleton
7. Spy called with to_s; Singleton is still the owner of to_s
8. Spy::Collection finds the Spy::Collection::Entry and raises AlreadySpiedError

**Comment by jbodah** [comment]
```rb
irb(main):009:0> Spy.on(Object, :to_s)
=> #<Spy::Instance:0x007f975abb92b8 @spied=Object, @original=#<Method: Class(Module)#to_s>, @visibility=:public, @conditional_filters=[], @before_callbacks=[], @after_callbacks=[], @around_procs=[], @call_history=[], @strategy=#<Spy::Instance::Strategy::Intercept:0x007f975abb9128 @spy=#<Spy::Instance:0x007f975abb92b8 ...>, @intercept_target=#<Class:Object>>>
irb(main):012:0> Spy.instance_variable_get(:@core).spy_collection
=> #<Spy::Collection:0x007f975abb9100 @store={"70144000875460|to_s|Method"=>#<struct Spy::Collection::Entry receiver=Object, msg=:to_s, method_type=Method>}>
irb(main):013:0> Spy.on(Object, :to_s)
=> #<Spy::Instance:0x007f975ab8e6f8 @spied=Object, @original=#<Method: Object.to_s>, @visibility=:public, @conditional_filters=[], @before_callbacks=[], @after_callbacks=[], @around_procs=[], @call_history=[], @strategy=#<Spy::Instance::Strategy::Wrap:0x007f975ab8e3b0 @spy=#<Spy::Instance:0x007f975ab8e6f8 ...>>>
irb(main):014:0> Spy.instance_variable_get(:@core).spy_collection
=> #<Spy::Collection:0x007f975abb9100 @store={"70144000875460|to_s|Method"=>#<struct Spy::Collection::Entry receiver=Object, msg=:to_s, method_type=Method>, "70144000875460|to_s|Object"=>#<struct Spy::Collection::Entry receiver=Object, msg=:to_s, method_type=Method>}>
```

**Comment by jbodah** [comment]
I think there's also an issue with spies still wrapping even though the instance exists due to the check being called after the instance is bound:

```rb
    def add_spy(spied, method)
      spy = Instance.new(spied, method)
      spy_collection << spy
      spy.start
    end
```

**Comment by jbodah** [comment]
Original issue seems like an edge case. Second issue that came up will be addressed in e2120b0753b688db211bc75a867174e1cbbf8013
